pkg/imgpkg/bundle/contents_test.go

```diff
@@ -0,0 +1,156 @@
+		subject := bundle.NewContents([]string{bundleDir}, nil, false)
+		imgTag, err := name.NewTag("my.registry.io/new-bundle:tag")
+		if err != nil {
+			t.Fatalf("failed to read tag: %s", err)
+		}
+
+		_, err = subject.Push(imgTag, fakeRegistry, fakeUI)
+		if err == nil {
+			t.Fatalf("expected test to fail, but it did not")
+		}
+		expectedError := "Expected image lock to not contain bundle reference: 'my.registry.io/bundle@sha256:703218c0465075f4425e58fac086e09e1de5c340b12976ab9eb8ad26615c3715'"
+		if err.Error() != expectedError {
+			t.Fatalf("Error message not expected\nExpected: %s\nGot: %s", expectedError, err.Error())
+		}
+	})
+
+	t.Run("when allowInnerBundles is true, push is successful", func(t *testing.T) {
+		subject := bundle.NewContents([]string{bundleDir}, nil, true)
+		imgTag, err := name.NewTag("my.registry.io/new-bundle:tag")
+		if err != nil {
+			t.Fatalf("failed to read tag: %s", err)
+		}
+
+		_, err = subject.Push(imgTag, fakeRegistry, fakeUI)
+		if err != nil {
+			t.Fatalf("not expecting push to fail: %s", err)
+		}
+	})
+}
+
+func TestNewContentsBundleWithImages(t *testing.T) {
+	imagesLockYAML := `---
+apiVersion: imgpkg.carvel.dev/v1alpha1
+kind: ImagesLock
+images:
+- image: my.registry.io/image1@sha256:703218c0465075f4425e58fac086e09e1de5c340b12976ab9eb8ad26615c3715
+`
+	fakeUI := &bundlefakes.FakeUI{}
+	fakeRegistry := &bundlefakes.FakeImagesMetadataWriter{}
+	bundleDir := createBundleDir(t, "inner-bundle", imagesLockYAML)
+	defer os.RemoveAll(bundleDir)
+	bundleImg := &fake.FakeImage{}
+	cfgFile := &v1.ConfigFile{
+		Config: v1.Config{},
+	}
+	bundleImg.ConfigFileReturns(cfgFile, nil)
+	fakeRegistry.ImageReturns(bundleImg, nil)
+
+	t.Run("when allowInnerBundles is false, push is successful", func(t *testing.T) {
+		subject := bundle.NewContents([]string{bundleDir}, nil, false)
+		imgTag, err := name.NewTag("my.registry.io/new-bundle:tag")
+		if err != nil {
+			t.Fatalf("failed to read tag: %s", err)
+		}
+
+		_, err = subject.Push(imgTag, fakeRegistry, fakeUI)
+		if err != nil {
+			t.Fatalf("not expecting push to fail: %s", err)
+		}
+	})
+
+	t.Run("when allowInnerBundles is true, push is successful", func(t *testing.T) {
+		subject := bundle.NewContents([]string{bundleDir}, nil, true)
+		imgTag, err := name.NewTag("my.registry.io/new-bundle:tag")
+		if err != nil {
+			t.Fatalf("failed to read tag: %s", err)
+		}
+
+		_, err = subject.Push(imgTag, fakeRegistry, fakeUI)
+		if err != nil {
+			t.Fatalf("not expecting push to fail: %s", err)
+		}
+	})
+}
+
+func createTempFolder(t *testing.T, prefix string) string {
+	t.Helper()
+	if prefix == "" {
+		prefix = "bundle"
+	}
+
+	rDir := filepath.Join(os.TempDir(), fmt.Sprintf("%s-%s", prefix, randString(8)))
+	err := os.MkdirAll(rDir, 0700)
+	if err != nil {
+		t.Fatalf("unable to create bundle folder: %s", err)
+	}
+	return rDir
+}
+
+func createBundleDir(t *testing.T, prefix string, iYml string) string {
```

Comment:
iYml -> imagesYAML?

Reply:
also saw `bundleDir := env.BundleFactory.CreateBundleDir(bundleYAML, imagesYAML)`. can we use that instead of adding new method?

Reply:
I thought about that but did not want to add a dependency between e2e tests and these unit tests. So for now my idea was to do some code duplication until we see what would be the best way to abstract this.
After reviewing the PR for the performance improvements I am starting to think that maybe we should move these functions to the package `test` so that it can serve all types of tests.

Nevertheless, I think that this change should be done by itself.

Reply:
ok that's fair since there is a unit tests vs e2e tests separation.